<h1><u><b>MULTI-LABEL CLASSIFICATION</b></u></h1>

The other approaches which could be used for other similar problem but is not suitable for this particular case are:
<ul>
    <li><b>Tensorflow_Ranking</b>: 
        This is the latest neaural network based ranking model for massive amount of data(which we pretty much dont have), training it on relatively smaller datasets can result in very poor results. This will give us a ranked order of the top n universities that the candidate should consider applying, but this is not justified as a candidate with very good score and academics may have greater number of options than a candidate with very less marks and academically weak, we can not provide same number of options for both the candidates.
        
   <li><b>Multi-Class Classification</b>:
        This will provide one output university that the candidate should consider applying to, this is not the right thing to do as th candidate does not apply to only one university and always likes more options, therfore this solution is nit accepted.
        
<li><b>Multi-Label Classification</b>:
    Therefore we have to adapt a more adaptive solution, which can display more results to a person who has person who good academics and relatively lesser and more feasible options to a person with weaker academic records. 

In [1]:
import collections
import pandas
from collections import defaultdict
import numpy as np 
import pandas as pd 
from sklearn.ensemble import RandomForestClassifier
from boruta import BorutaPy
from numpy.random import permutation
import math

In [2]:
data=pd.read_csv('data.csv')

In [3]:
data=data.drop(['Unnamed: 0'],1)

In [4]:
data

userName  toeflScore   greV   greQ  greA    termAndYear  \
0             ninny.rgs       101.0  430.0  740.0   4.5    Fall - 2011   
1             ninny.rgs       101.0  159.0  159.0   4.5    Fall - 2011   
2            aashayshah        88.0  147.0  164.0   2.5  Spring - 2016   
3           cjaganmohan       102.0  140.0  148.0   3.0    Fall - 2013   
4                pvskmz       115.0  154.0  163.0   3.0    Fall - 2013   
5               anu_661       105.0  154.0  163.0   3.0    Fall - 2011   
6         nachiappan123       100.0  157.0  163.0   3.0    Fall - 2015   
7        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
8        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
9        sam_the_killer        89.0  370.0  690.0   4.0    Fall - 2007   
10       sam_the_killer        89.0  154.0  163.0   4.0    Fall - 2007   
11           dhriti1991       101.0  154.0  155.0   3.5    Fall - 2014   
12           dhriti1991       101.0  157.0  163.0   3.5    Fall - 2014   
13          hemalpatel1        85.0  162.0  163.0   2.5  Spring - 2011   
14              kranth1       100.0  157.0  163.0   3.5    Fall - 2013   
15            johncena3        96.0  165.0  160.0   3.0  Spring - 2011   
16          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
17          abhi111cute        95.0  154.0  160.0   3.0    Fall - 2011   
18              kranth1       100.0  162.0  156.0   3.5    Fall - 2013   
19          abhi111cute        95.0  153.0  154.0   3.0    Fall - 2011   
20            nathnandy        95.0  162.0  156.0   3.0    Fall - 2015   
21            MISfinder       107.0  162.0  156.0   5.0    Fall - 2014   
22           praneeth4u        90.0  162.0  156.0   3.0    Fall - 2014   
23           praneeth4u        90.0  141.0  159.0   3.0    Fall - 2014   
24          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
25                  s29       102.0  630.0  780.0   3.0    Fall - 2010   
26          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
27          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
28          supreeth324       107.0  156.0  166.0   3.5    Fall - 2012   
29     chandra.chowdary       102.0  146.0  164.0   3.0    Fall - 2010   
...                 ...         ...    ...    ...   ...            ...   
21285     aviral_sawani       109.0  158.0  167.0   4.0    Fall - 2013   
21286        sg09161988       100.0  157.0  150.0   3.0    Fall - 2014   
21287      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21288            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21289   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21290      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21291   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21292   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21293    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21294      swatiebansal       105.0  147.0  170.0   3.5    Fall - 2014   
21295            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21296     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21297      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21298       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21299       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21300     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21301        ishanpahwa       101.0  149.0  170.0   3.5    Fall - 2014   
21302    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21303         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21304         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21305            raj.65       112.0  156.0  170.0   3.5    Fall - 2014   
21306         airyuanye       102.0  680.0  800.0   3.0    Fall - 2012   
21307         airyu

In [5]:
one_hot = pd.get_dummies(data['univName'])

In [6]:
data = data.drop('univName',axis = 1)

In [7]:
data = data.join(one_hot)

In [8]:
data

userName  toeflScore   greV   greQ  greA    termAndYear  \
0             ninny.rgs       101.0  430.0  740.0   4.5    Fall - 2011   
1             ninny.rgs       101.0  159.0  159.0   4.5    Fall - 2011   
2            aashayshah        88.0  147.0  164.0   2.5  Spring - 2016   
3           cjaganmohan       102.0  140.0  148.0   3.0    Fall - 2013   
4                pvskmz       115.0  154.0  163.0   3.0    Fall - 2013   
5               anu_661       105.0  154.0  163.0   3.0    Fall - 2011   
6         nachiappan123       100.0  157.0  163.0   3.0    Fall - 2015   
7        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
8        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
9        sam_the_killer        89.0  370.0  690.0   4.0    Fall - 2007   
10       sam_the_killer        89.0  154.0  163.0   4.0    Fall - 2007   
11           dhriti1991       101.0  154.0  155.0   3.5    Fall - 2014   
12           dhriti1991       101.0  157.0  163.0   3.5    Fall - 2014   
13          hemalpatel1        85.0  162.0  163.0   2.5  Spring - 2011   
14              kranth1       100.0  157.0  163.0   3.5    Fall - 2013   
15            johncena3        96.0  165.0  160.0   3.0  Spring - 2011   
16          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
17          abhi111cute        95.0  154.0  160.0   3.0    Fall - 2011   
18              kranth1       100.0  162.0  156.0   3.5    Fall - 2013   
19          abhi111cute        95.0  153.0  154.0   3.0    Fall - 2011   
20            nathnandy        95.0  162.0  156.0   3.0    Fall - 2015   
21            MISfinder       107.0  162.0  156.0   5.0    Fall - 2014   
22           praneeth4u        90.0  162.0  156.0   3.0    Fall - 2014   
23           praneeth4u        90.0  141.0  159.0   3.0    Fall - 2014   
24          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
25                  s29       102.0  630.0  780.0   3.0    Fall - 2010   
26          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
27          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
28          supreeth324       107.0  156.0  166.0   3.5    Fall - 2012   
29     chandra.chowdary       102.0  146.0  164.0   3.0    Fall - 2010   
...                 ...         ...    ...    ...   ...            ...   
21285     aviral_sawani       109.0  158.0  167.0   4.0    Fall - 2013   
21286        sg09161988       100.0  157.0  150.0   3.0    Fall - 2014   
21287      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21288            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21289   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21290      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21291   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21292   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21293    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21294      swatiebansal       105.0  147.0  170.0   3.5    Fall - 2014   
21295            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21296     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21297      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21298       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21299       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21300     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21301        ishanpahwa       101.0  149.0  170.0   3.5    Fall - 2014   
21302    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21303         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21304         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21305            raj.65       112.0  156.0  170.0   3.5    Fall - 2014   
21306         airyuanye       102.0  680.0  800.0   3.0    Fall - 2012   
21307         airyu

In [9]:
temp=data.copy()

In [10]:
comp=data.copy()

In [11]:
temp

userName  toeflScore   greV   greQ  greA    termAndYear  \
0             ninny.rgs       101.0  430.0  740.0   4.5    Fall - 2011   
1             ninny.rgs       101.0  159.0  159.0   4.5    Fall - 2011   
2            aashayshah        88.0  147.0  164.0   2.5  Spring - 2016   
3           cjaganmohan       102.0  140.0  148.0   3.0    Fall - 2013   
4                pvskmz       115.0  154.0  163.0   3.0    Fall - 2013   
5               anu_661       105.0  154.0  163.0   3.0    Fall - 2011   
6         nachiappan123       100.0  157.0  163.0   3.0    Fall - 2015   
7        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
8        username123456        95.0  143.0  162.0   3.0  Spring - 2014   
9        sam_the_killer        89.0  370.0  690.0   4.0    Fall - 2007   
10       sam_the_killer        89.0  154.0  163.0   4.0    Fall - 2007   
11           dhriti1991       101.0  154.0  155.0   3.5    Fall - 2014   
12           dhriti1991       101.0  157.0  163.0   3.5    Fall - 2014   
13          hemalpatel1        85.0  162.0  163.0   2.5  Spring - 2011   
14              kranth1       100.0  157.0  163.0   3.5    Fall - 2013   
15            johncena3        96.0  165.0  160.0   3.0  Spring - 2011   
16          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
17          abhi111cute        95.0  154.0  160.0   3.0    Fall - 2011   
18              kranth1       100.0  162.0  156.0   3.5    Fall - 2013   
19          abhi111cute        95.0  153.0  154.0   3.0    Fall - 2011   
20            nathnandy        95.0  162.0  156.0   3.0    Fall - 2015   
21            MISfinder       107.0  162.0  156.0   5.0    Fall - 2014   
22           praneeth4u        90.0  162.0  156.0   3.0    Fall - 2014   
23           praneeth4u        90.0  141.0  159.0   3.0    Fall - 2014   
24          abhi111cute        95.0  520.0  780.0   3.0    Fall - 2011   
25                  s29       102.0  630.0  780.0   3.0    Fall - 2010   
26          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
27          supreeth324       107.0  520.0  710.0   3.5    Fall - 2012   
28          supreeth324       107.0  156.0  166.0   3.5    Fall - 2012   
29     chandra.chowdary       102.0  146.0  164.0   3.0    Fall - 2010   
...                 ...         ...    ...    ...   ...            ...   
21285     aviral_sawani       109.0  158.0  167.0   4.0    Fall - 2013   
21286        sg09161988       100.0  157.0  150.0   3.0    Fall - 2014   
21287      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21288            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21289   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21290      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21291   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21292   neerajyadav.ggn       107.0  154.0  162.0   4.0    Fall - 2012   
21293    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21294      swatiebansal       105.0  147.0  170.0   3.5    Fall - 2014   
21295            ajaysa       108.0  670.0  800.0   3.5    Fall - 2012   
21296     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21297      kartikaygarg       111.0  158.0  170.0   3.5    Fall - 2015   
21298       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21299       surbhidec24       113.0  155.0  166.0   3.5    Fall - 2014   
21300     ankita.kak101        97.0  157.0  163.0   3.0    Fall - 2014   
21301        ishanpahwa       101.0  149.0  170.0   3.5    Fall - 2014   
21302    nitishbhatia11       106.0  149.0  163.0   4.0    Fall - 2014   
21303         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21304         shammisrz        99.0  620.0  740.0   3.0    Fall - 2012   
21305            raj.65       112.0  156.0  170.0   3.5    Fall - 2014   
21306         airyuanye       102.0  680.0  800.0   3.0    Fall - 2012   
21307         airyu

In [12]:
temp.groupby(['userName'])

In [13]:
comp=comp.groupby(['userName']).max()

In [14]:
comp

toeflScore   greV   greQ  greA    termAndYear  \
userName                                                             
! Superman                118.0  165.0  169.0   5.5    Fall - 2014   
!mpulse                    95.0  140.0  166.0   3.0    Fall - 2013   
$4R4                      109.0  145.0  168.0   3.0    Fall - 2015   
$aumil                    109.0  610.0  750.0   3.0    Fall - 2013   
(:)_(:)                   103.0  150.0  166.0   3.0    Fall - 2013   
++ZeuS++                  110.0  660.0  760.0   4.0    Fall - 2011   
-=[prat17]=-               85.0  152.0  161.0   4.5    Fall - 2012   
-KaaL-                     97.0  142.0  157.0   3.5    Fall - 2012   
-MK-                       99.0  151.0  159.0   3.5    Fall - 2015   
007.soham@gmail.com       104.0  164.0  166.0   3.0    Fall - 2012   
007happa                  104.0  152.0  161.0   3.0    Fall - 2013   
0wn3d                     106.0  160.0  166.0   3.5    Fall - 2011   
100GODS                   105.0  570.0  740.0   3.5    Fall - 2015   
10sojniv                  102.0  152.0  160.0   3.0    Fall - 2015   
110kar3685                115.0  162.0  166.0   4.5    Fall - 2011   
12-Dec                    103.0  145.0  157.0   3.0    Fall - 2012   
123maela                  104.0  170.0  141.0   3.0    Fall - 2015   
13supriya.jain            109.0  150.0  160.0   3.0    Fall - 2013   
13vidya                   101.0  152.0  165.0   3.5    Fall - 2013   
1413                      112.0  157.0  165.0   3.5  Spring - 2015   
15gre                     107.0  144.0  158.0   4.0    Fall - 2012   
1987frank                 105.0  153.0  170.0   3.5  Spring - 2014   
198921                    105.0  150.0  164.0   3.5    Fall - 2012   
1990                       95.0  165.0  164.0   3.0    Fall - 2015   
19arjun89                 108.0  158.0  164.0   4.0    Fall - 2014   
1q2w3e4r                  108.0  410.0  790.0   3.5  Summer - 2013   
2-Feb                     116.0  720.0  800.0   4.0    Fall - 2012   
200fever                  110.0  153.0  157.0   3.5  Spring - 2015   
2013_July                  98.0  166.0  166.0   3.0    Fall - 2014   
221B                      119.0  164.0  168.0   4.0    Fall - 2015   
...                         ...    ...    ...   ...            ...   
zab.zabid                 100.0   32.0   49.0   4.5    Fall - 2012   
zackriyaniyas              92.0  420.0  720.0   3.5  Spring - 2012   
zakoraio                   91.0  155.0  164.0   3.5    Fall - 2014   
zambro                    106.0  163.0  159.0   4.0    Fall - 2012   
zapp11                     98.0  151.0  152.0   3.0    Fall - 2014   
zaraki                    104.0  150.0  164.0   3.0    Fall - 2013   
zarthon                   110.0  164.0  166.0   3.0    Fall - 2013   
zealot295                 110.0  580.0  800.0   4.0    Fall - 2012   
zebanaqvi                 108.0  159.0  163.0   4.0    Fall - 2012   
zeevee                    111.0  660.0  730.0   3.0  Spring - 2012   
zeferosri                 108.0  154.0  163.0   3.5    Fall - 2015   
zenga                      92.0  145.0  161.0   3.0    Fall - 2012   
zenith1990                106.0  151.0  159.0   3.5    Fall - 2012   
zenith48                   96.0  165.0  166.0   3.0    Fall - 2012   
zeph31                    108.0  158.0  153.0   3.5    Fall - 2012   
zephyr_dce                113.0  670.0  800.0   4.5    Fall - 2011   
zeusaka                    96.0  151.0  158.0   3.0    Fall - 2014   
zidanefaraday              99.0  470.0  760.0   3.5    Fall - 2011   
zion                      109.0  680.0  800.0   3.5  Spring - 2013   
zizou007                  113.0  159.0  163.0   3.5  Spring - 2014   
zohaib16                  104.0  460.0  740.0   4.0    Fall - 2011   
zoheb.haider              110.0  154.0  163.0   5.0    Fall - 2013   
zohebtahir                102.0  149.0  155.0   3.0  Spring - 2013   
zoomsthru                 290.0  730.0  800.0   5.5    Fall - 2006   
zrini1901                  92.0

In [15]:
len(comp.columns)

50

In [16]:
from sklearn import preprocessing
number = preprocessing.LabelEncoder()
comp['termAndYear'] = number.fit_transform(comp.termAndYear)
comp['ugCollege'] = number.fit_transform(comp.ugCollege)


In [17]:
comp

toeflScore   greV   greQ  greA  termAndYear  ugCollege  \
userName                                                                      
! Superman                118.0  165.0  169.0   5.5           20        156   
!mpulse                    95.0  140.0  166.0   3.0           17        565   
$4R4                      109.0  145.0  168.0   3.0           21         66   
$aumil                    109.0  610.0  750.0   3.0           17        559   
(:)_(:)                   103.0  150.0  166.0   3.0           17         43   
++ZeuS++                  110.0  660.0  760.0   4.0           14        524   
-=[prat17]=-               85.0  152.0  161.0   4.5           16        831   
-KaaL-                     97.0  142.0  157.0   3.5           16        524   
-MK-                       99.0  151.0  159.0   3.5           21        938   
007.soham@gmail.com       104.0  164.0  166.0   3.0           16        675   
007happa                  104.0  152.0  161.0   3.0           17        298   
0wn3d                     106.0  160.0  166.0   3.5           14        708   
100GODS                   105.0  570.0  740.0   3.5           21        486   
10sojniv                  102.0  152.0  160.0   3.0           21        565   
110kar3685                115.0  162.0  166.0   4.5           14       1066   
12-Dec                    103.0  145.0  157.0   3.0           16        145   
123maela                  104.0  170.0  141.0   3.0           21        869   
13supriya.jain            109.0  150.0  160.0   3.0           17        872   
13vidya                   101.0  152.0  165.0   3.5           17        565   
1413                      112.0  157.0  165.0   3.5           38        965   
15gre                     107.0  144.0  158.0   4.0           16        111   
1987frank                 105.0  153.0  170.0   3.5           37        798   
198921                    105.0  150.0  164.0   3.5           16        565   
1990                       95.0  165.0  164.0   3.0           21        337   
19arjun89                 108.0  158.0  164.0   4.0           20        433   
1q2w3e4r                  108.0  410.0  790.0   3.5           43        248   
2-Feb                     116.0  720.0  800.0   4.0           16        298   
200fever                  110.0  153.0  157.0   3.5           38        565   
2013_July                  98.0  166.0  166.0   3.0           20        926   
221B                      119.0  164.0  168.0   4.0           21        524   
...                         ...    ...    ...   ...          ...        ...   
zab.zabid                 100.0   32.0   49.0   4.5           16        852   
zackriyaniyas              92.0  420.0  720.0   3.5           35        797   
zakoraio                   91.0  155.0  164.0   3.5           20        916   
zambro                    106.0  163.0  159.0   4.0           16        635   
zapp11                     98.0  151.0  152.0   3.0           20        524   
zaraki                    104.0  150.0  164.0   3.0           17        901   
zarthon                   110.0  164.0  166.0   3.0           17        822   
zealot295                 110.0  580.0  800.0   4.0           16         47   
zebanaqvi                 108.0  159.0  163.0   4.0           16        841   
zeevee                    111.0  660.0  730.0   3.0           35        945   
zeferosri                 108.0  154.0  163.0   3.5           21        142   
zenga                      92.0  145.0  161.0   3.0           16        565   
zenith1990                106.0  151.0  159.0   3.5           16        297   
zenith48                   96.0  165.0  166.0   3.0           16        687   
zeph31                    108.0  158.0  153.0   3.5           16        635   
zephyr_dce                113.0  670.0  800.0   4.5           14       1109   
zeusaka                    96.0  151.0  158.0   3.0           20         85   
zidanefaraday              99.0  470.0  760.0   3.5           14       

In [18]:
x = comp.iloc[:,0:7] 
y = comp.iloc[:,7:50]

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size = 0.30, random_state = 0)

In [20]:
X_test

toeflScore   greV   greQ  greA  termAndYear  ugCollege  \
userName                                                                      
prasanna1433               93.0  144.0  161.0   3.0           20        387   
kirti8192                 113.0  164.0  164.0   4.0           17       1016   
GT-R                      112.0  157.0  155.0   4.0           16        737   
karthikraoa               114.0  640.0  800.0   3.5           14         66   
priyatham                  99.0  460.0  800.0   3.5           14        839   
yuvarajchiyyodu           104.0  167.0  170.0   3.0           17        337   
tourniquet                111.0  162.0  161.0   4.5           17        202   
sahildude                 106.0  164.0  158.0   3.0           17        913   
abhishekjoshi             117.0  164.0  170.0   5.5           21       1155   
navneet_reddevil18         93.0  148.0  157.0   3.0           21        754   
clnsdragon                104.0  380.0  790.0   3.0           14        312   
rajat.singh_203            93.0  143.0  156.0   3.5           17        615   
shraddhasb                111.0  630.0  800.0   3.0           14         47   
Harsha Govada              90.0  162.0  163.0   2.5           37        938   
ruchi.smiley              106.0  159.0  159.0   3.5           20        842   
Chinnu_A                  100.0  146.0  162.0   3.0           20       1027   
adream                    116.0  160.0  157.0   4.5           14        653   
pratikgandhi8288           93.0  151.0  158.0   3.0           39        553   
kedaritheja               113.0  610.0  790.0   3.5           16       1026   
rupa387                   109.0  590.0  790.0   3.5           14        724   
black_mamba               102.0  520.0  790.0   3.5           14        693   
asvin_1989                110.0  500.0  800.0   3.5           17        297   
vivek                     290.0  600.0  800.0   5.0            9        781   
nirany                     94.0  410.0  700.0   3.0           16        549   
007.soham@gmail.com       104.0  164.0  166.0   3.0           16        675   
mtkc                      113.0  570.0  780.0   4.0           14        102   
pavan009                   98.0  151.0  166.0   3.5           20        607   
avenkat10                  90.0  380.0  800.0   3.5           16        524   
Traveller99               108.0  154.0  166.0   2.5            3        690   
rahul8805                 108.0  410.0  730.0   4.0           14       1027   
...                         ...    ...    ...   ...          ...        ...   
dey.archit                108.0  153.0  162.0   3.5           21        565   
dinesh79                  104.0  168.0  160.0   3.0           20        607   
Gauty                      99.0  150.0  156.0   3.0           20        307   
HedgeHog                  101.0  510.0  800.0   3.5           16        546   
gaurp                      99.0  360.0  780.0   3.0           16        111   
chirayu.diwan09           110.0  680.0  780.0   4.0           12        461   
utsumneo                  100.0  162.0  168.0   3.5           16        288   
Nebelwerfer                97.0  500.0  800.0   4.0           14        614   
srik506                    90.0  390.0  800.0   3.5           14        686   
nikhil_ramam              104.0  430.0  800.0   4.0           14        979   
dilarasemerci             100.0  142.0  168.0   3.0           16        728   
venkatj                    97.0  159.0  166.0   3.0           16        645   
dfidt                      97.0  148.0  161.0   3.0           21        599   
rohit87                   104.0   34.0   49.0   5.0            0        477   
ashesh29nov                96.0  149.0  158.0   3.0           17        337   
rohit.panhale.7           104.0  530.0  800.0   3.5           16        579   
sirimuvva                 103.0  570.0  800.0   3.5           16         51   
drugandhi                 114.0  620.0  800.0   4.0           16       

In [21]:
import scipy.sparse as sp
X_train = sp.csr_matrix(X_train)
Y_train = sp.csr_matrix(Y_train)
X_test = sp.csr_matrix(X_test)
Y_test = sp.csr_matrix(Y_test)

In [22]:
sam_x = comp.iloc[:,0:7] 
sam_y = comp.iloc[:,7:50]

In [23]:
sam_x

toeflScore   greV   greQ  greA  termAndYear  ugCollege  \
userName                                                                      
! Superman                118.0  165.0  169.0   5.5           20        156   
!mpulse                    95.0  140.0  166.0   3.0           17        565   
$4R4                      109.0  145.0  168.0   3.0           21         66   
$aumil                    109.0  610.0  750.0   3.0           17        559   
(:)_(:)                   103.0  150.0  166.0   3.0           17         43   
++ZeuS++                  110.0  660.0  760.0   4.0           14        524   
-=[prat17]=-               85.0  152.0  161.0   4.5           16        831   
-KaaL-                     97.0  142.0  157.0   3.5           16        524   
-MK-                       99.0  151.0  159.0   3.5           21        938   
007.soham@gmail.com       104.0  164.0  166.0   3.0           16        675   
007happa                  104.0  152.0  161.0   3.0           17        298   
0wn3d                     106.0  160.0  166.0   3.5           14        708   
100GODS                   105.0  570.0  740.0   3.5           21        486   
10sojniv                  102.0  152.0  160.0   3.0           21        565   
110kar3685                115.0  162.0  166.0   4.5           14       1066   
12-Dec                    103.0  145.0  157.0   3.0           16        145   
123maela                  104.0  170.0  141.0   3.0           21        869   
13supriya.jain            109.0  150.0  160.0   3.0           17        872   
13vidya                   101.0  152.0  165.0   3.5           17        565   
1413                      112.0  157.0  165.0   3.5           38        965   
15gre                     107.0  144.0  158.0   4.0           16        111   
1987frank                 105.0  153.0  170.0   3.5           37        798   
198921                    105.0  150.0  164.0   3.5           16        565   
1990                       95.0  165.0  164.0   3.0           21        337   
19arjun89                 108.0  158.0  164.0   4.0           20        433   
1q2w3e4r                  108.0  410.0  790.0   3.5           43        248   
2-Feb                     116.0  720.0  800.0   4.0           16        298   
200fever                  110.0  153.0  157.0   3.5           38        565   
2013_July                  98.0  166.0  166.0   3.0           20        926   
221B                      119.0  164.0  168.0   4.0           21        524   
...                         ...    ...    ...   ...          ...        ...   
zab.zabid                 100.0   32.0   49.0   4.5           16        852   
zackriyaniyas              92.0  420.0  720.0   3.5           35        797   
zakoraio                   91.0  155.0  164.0   3.5           20        916   
zambro                    106.0  163.0  159.0   4.0           16        635   
zapp11                     98.0  151.0  152.0   3.0           20        524   
zaraki                    104.0  150.0  164.0   3.0           17        901   
zarthon                   110.0  164.0  166.0   3.0           17        822   
zealot295                 110.0  580.0  800.0   4.0           16         47   
zebanaqvi                 108.0  159.0  163.0   4.0           16        841   
zeevee                    111.0  660.0  730.0   3.0           35        945   
zeferosri                 108.0  154.0  163.0   3.5           21        142   
zenga                      92.0  145.0  161.0   3.0           16        565   
zenith1990                106.0  151.0  159.0   3.5           16        297   
zenith48                   96.0  165.0  166.0   3.0           16        687   
zeph31                    108.0  158.0  153.0   3.5           16        635   
zephyr_dce                113.0  670.0  800.0   4.5           14       1109   
zeusaka                    96.0  151.0  158.0   3.0           20         85   
zidanefaraday              99.0  470.0  760.0   3.5           14       

In [24]:
from sklearn.model_selection import train_test_split
samxtrain, samxtest, samytrain, samytest = train_test_split(sam_x, sam_y, test_size = 0.30, random_state = 0)

In [98]:
# using Label Powerset
from skmultilearn.problem_transform import LabelPowerset
from sklearn.naive_bayes import GaussianNB

# initialize Label Powerset multi-label classifier
# with a gaussian naive bayes base classifier
classifier = LabelPowerset(GaussianNB())

# train
classifier.fit(samxtrain, samytrain)

# predict
predictions = classifier.predict(samxtest)
from sklearn.metrics import roc_auc_score
roc_auc_score(samytest,predictions.toarray(),average='samples')

0.5554965551518193

In [109]:
# using Label Powerset
from skmultilearn.problem_transform import LabelPowerset
from sklearn.naive_bayes import GaussianNB

# initialize Label Powerset multi-label classifier
# with a gaussian naive bayes base classifier
classifier = LabelPowerset(GaussianNB())

# train
classifier.fit(samxtrain, samytrain)

# predict
predictions = classifier.predict(samxtrain)
from sklearn.metrics import roc_auc_score
roc_auc_score(samytrain,predictions.toarray(),average='samples')

0.6824250275661522

In [112]:
from skmultilearn.adapt import MLkNN

classifier = MLkNN(k=10,s=0.5)

# train
classifier.fit(X_train, Y_train)

# predict
predictions = classifier.predict(X_test)
from sklearn.metrics import roc_auc_score
roc_auc_score(Y_test.toarray(),predictions.toarray(),average='samples')

0.5098644325237893

In [111]:
from skmultilearn.adapt import MLkNN

classifier = MLkNN(k=10,s=0.5)

# train
classifier.fit(X_train, Y_train)

# predict
predictions = classifier.predict(X_train)
from sklearn.metrics import roc_auc_score
roc_auc_score(Y_train.toarray(),predictions.toarray(),average='samples')

0.5187412047203723

<h1><u>PERFORMANCE</u></h1> 

The approaches used in the above multi-label classification are:

<h2><b>Problem Transformation:</b></h2> In thos method we try to transform our multi-label problem into single-label problems. One of the ways to achieve this is by using label powerset with Naive Bayes classification algorithm. <b>The train data accuracy is 68.2% and the test data accuracy is 55.5%, which is higher than the accuracy specified in the report provided.</b>

<h2><b>Adapted Algorithm:</b></h2> This method suggests adapting the algorithm to directly perform multi-label classification, rather than transforming the problem into different subsets of problems.One of the ways to implement this is ny using multi-label version of kNN (MLkNN). the train data accuarcy is 51.8% and test data accuracy is 50.9%

Ref:https://www.analyticsvidhya.com/blog/2017/08/introduction-to-multi-label-classification/